### PR TITLE
BAEL-3223: Excluding URLs from a Spring Filter

### PR DESCRIPTION
**Pull request by tapankavasthi**
- Add source code for a tutorial on Excluding URLs for a Filter in a Spring Web Application:
   - Sample Spring Web Application using Custom Filters
   - Custom Logic in Filters to Exclude URLs
- Reference: [BAEL-3223](http://jira.baeldung.com/browse/BAEL-3223)

### CODE REVIEW

File: spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/controller/FAQController.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/faq")
+public class FAQController {
+
+    private FAQService faqService;
+
+    @Autowired
+    public FAQController(FAQService faqService) {
```
Comment (rpvilao) [inline spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/controller/FAQController.java]: if we are doing constructor injection let's make the var final

File: spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/controller/FAQController.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        this.faqService = faqService;
+    }
+
+    @RequestMapping("/helpline")
```
Comment (rpvilao) [inline spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/controller/FAQController.java]: It's a good practice to restrict the mapping to the method, in this case, it seems it's just a get

File: spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/controller/Ping.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/health")
+public class Ping {
+
+    @RequestMapping
```
Comment (rpvilao) [inline spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/controller/Ping.java]: same

File: spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/filter/HeaderContextFilter.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import org.springframework.core.annotation.Order;
+
+import javax.servlet.*;
+import javax.servlet.http.HttpServletRequest;
+import java.io.IOException;
+
+@Order(1)
+public class HeaderContextFilter implements Filter {
+
+	@Override
```
Comment (rpvilao) [inline spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/filter/HeaderContextFilter.java]: fix

File: spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/filter/LogFilter.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.core.annotation.Order;
+
+import javax.servlet.*;
+import javax.servlet.http.HttpServletRequest;
+import java.io.IOException;
+
+//@Component
```
Comment (rpvilao) [inline spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/filter/LogFilter.java]: if not needed let's remove

File: spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/filter/LogFilter.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    public void doFilter(
+            ServletRequest request,
+            ServletResponse response,
+            FilterChain chain) throws IOException, ServletException {
+
+        String path = ((HttpServletRequest) request).getRequestURI();
+        String contentType = request.getContentType();
+        logger.info("Path logged:" + path);
+        if (!path.endsWith("health")) {
+            logger.info("Mime type of request body is:" + contentType);
```
Comment (rpvilao) [inline spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/filter/LogFilter.java]: let's use the logging features .ingo("... {}", var1, ...)

File: spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/filter/HeaderContextFilter.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    public void doFilter(
+            ServletRequest request,
+            ServletResponse response,
+            FilterChain chain) throws IOException, ServletException {
+
+        String countryCodeHeader = ((HttpServletRequest) request).getHeader("COUNTRY_CODE");
+        HeaderContextProvider.setContext(countryCodeHeader);
+        chain.doFilter(request, response);
```
Comment (rpvilao) [inline spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/filter/HeaderContextFilter.java]: I am not totally sure but shouldn't that be protected in such a way:
HeaderContextProvider.setContext(countryCodeHeader);
  try {
    chain.doFilter(request, response);
} finally {
    HeaderContextProvider.unsetContext();
}

File: spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/filter/HeaderContextFilter.java (hunk 4 above)
Comment (tapankavasthi) [reply spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/filter/HeaderContextFilter.java]: ^ @rpvilao: I didn't understand what needs to be fixed here. Though, I've made changes in the same piece of code as per your other comments.
Please let me know if it still needs any change.

File: spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/filter/LogFilter.java (hunk 5 above)
Comment (tapankavasthi) [reply spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/filter/LogFilter.java]: Done 🙂

File: spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/filter/LogFilter.java (hunk 6 above)
Comment (tapankavasthi) [reply spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/filter/LogFilter.java]: Done 🙂

File: spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/filter/HeaderContextFilter.java (hunk 7 above)
Comment (tapankavasthi) [reply spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/filter/HeaderContextFilter.java]: Done 🙂

File: spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/controller/Ping.java (hunk 3 above)
Comment (tapankavasthi) [reply spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/controller/Ping.java]: Done 🙂

File: spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/controller/FAQController.java (hunk 2 above)
Comment (tapankavasthi) [reply spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/controller/FAQController.java]: Done 🙂

File: spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/controller/FAQController.java (hunk 1 above)
Comment (tapankavasthi) [reply spring-mvc-simple-2/src/main/java/com/baeldung/exclude_urls_filter/controller/FAQController.java]: Done 🙂

**Pull request merged**